### DISCUSSION TITLE
Default Feature Set

### DISCUSSION THREAD

**Discussion by exectails** [NoCode-NoLife/melia#270] [Discussions]
In some form or another, this has been repeatedly talked about on Discord, in issues, and discussions, but we didn't have a dedicated topic or official discussion about which version of ToS we're really targeting yet, so I'd like open one.

The game has gone through multiple iterations and has also always received regular updates that added, removed, or changed various aspects of the game. Examples include how certain skills work, which classes learn specific skills, how much damage certain attacks do, what kinds of equipment are available, and so on. One more example that we just implemented is active blocking, which once nullified damage entirely, but was later changed to only lower it. There's a _lot_ of big and small changes like this, and realistically we won't be able to ever implement all of them. Instead, there will always be one default state and the potential to enable or disable certain features, like enabling/disabling nullifying blocks. What's not entirely clear is what the default should be.

Many players prefer the classic ToS, often from around the time of its release, and there have been voices among our users that called for this to be the default. For combat to not be as easy as it is nowadays, for using the original class tree, for using circles, and many other things. The reason is simple: That's what we all want, so it would be pointless to even implement the newer game. While this argument makes a certain amount of sense, there are problems with it.

One of these is that the game has gone through so many iterations, that are so badly documented, that it would be nigh impossible to point to a single version and say, that's our point of reference. This muddies the waters in various ways. Not only will there never be that one version for us to look at during development, but users may also be confused because the default behavior might not match the specific behavior they had in mind, that might've been official at some point. Not to mention that there will always be features we'd have to improvise, because they're no longer testable, or perhaps didn't even exist in older versions. The end result of going this route will effectively always be some mishmash of various versions, essentially boiling down to whatever we, the developers, deemed best. Melia would become its own thing, removed from ToS, and any developers or players would simply have to deal with it, learning the default state over time. Now whether this is actually a problem or perhaps even desired, I'm honestly not sure, as there are pros and cons to be considered. However, there is also one big alternative that has many advantages.

It is my personal opinion that the best default state for any server emulator for a game that is in active development _probably_ is the current feature set. That's the one version all users have easy access to, it's what we can all actively test against, and it's what we can point towards and say, this is what we're implementing. If there is ever any question about how something is supposed to work, that's where you can check, regardless of your role in the community. If you ignore whether you like that version or not, it's just what makes the most sense, and it's usually not very difficult to toggle certain features, like we already are via our feature db. It also has the nice side-effect that we're really researching all aspects of the game, which is great for preservation sake.

I'd like to note that I am most definitely not a fan of the current version either though, and I'd never want to play it, which is why I understand any critique in this regard. It would most certainly be faster to implement features however we see fit, matching the version we preferred, and maybe even ignore any toggles in any direction unless we find them useful. At least if you ignore any potential arguments. Perhaps active blocking _should_ nullify damage by default, and _if_ we decide to add an option for the alternative, well, that's a bonus for the user. Melia wouldn't technically have to be a black box either. We could create a reference document that details our decisions, which features work how, and why.

But the path of least resistance in my opinion is simply to use the latest version as the primary reference. Like I said, that's really the only reliable reference we have, and that's the way I would personally go, while adding options to revert undesirable changes. And since we're in the fortunate position of being able to modify even the UI from the server-side, we'll be able to go far with this tactic. For example, that's what allows us to make stat points usable again, even though this feature was removed with ReBuild. We will be able to revert 90% of the game back to what it was like at release, albeit with a bit of a detour as we're also implementing the latest version as the default.

Though I also want to be clear about how these features are going to be implemented. When I say "let's do it like officials," I don't mean we have to implement premium classes as premium only for example, with tokens and buying premium cash, and the likes. What I'm talking about is how the classes function, not necessarily how you obtain them. And of course I would add some option to get them another, free way. It's really more about the functionality at the core. About whether block nullifies, not whether you need a premium shield for it and how you obtain it.

With all that being said, I'm open to other opinions. I suspect that it would be rather difficult to reach a consensus about how the game should function, given the number of changes made to it over the years, and the different opinions on what this game should be like, but it is a discussion worth having, and with this I at least explained my position on this matter.

**Comment by kenedos** [comment]
I agree fully that the current version should be the standard. It's what it's easily accessible and doesn't require specific client version distribution from our side (which would be illegal anyways).

Anything before the current version should be viewed as optional **and custom**. It is already hard enough to verify up-to-date information, let's not waste time discussing how a certain skill or feature worked back in 2015. If it's old - it should be treated as custom, and thus, does not even require credible oficial sources. It should be entirely up to the developer who's implementing it if they want to add that extra optional feature or not. However, the baseline and standard implementation should always be based on the newest version.

**Reply by exectails** [reply]
Considering old and unverifiable information custom is an interesting point of view. Makes a lot of sense.

**Reply by Adephx** [reply]
I was sure that we can revert most of client back with LUA scripts alone? I don't think anyone wants to actually use the old rusty client which crashed every 20 seconds. Everything that can't be reverted back with LUA scripts, we can most probably revert back with a client mod.

**Comment by felipecastagnarodecarvalho** [comment]
I feel like the project should follow the official core functionalities.
I am wondering if we gonna implement the story line quests since it makes part of the core of the officials.
If we choose to not insert quests it would be very hard to keep up with other functionalities since most of the game progress
(early/mid game) depends on it.

Gameplay concepts that I like on the official:
 - Main quests are rewarding enough.
 - Receiving equipments at the start that you gonna take until the late game is awesome and the progress feels smooth.
 - Rankings adds a good competitive spirit.
 - Items trades restrictions are good enough while some important items must be acquired by yourself.
 - PVE on the late game is fun if you put enough gameplay time into the game.
 - Being able to switch classes quickly
 - Being able to save character skills
 
 Gameplay concepts that I hate on officials:
  - There are too many main quests and they often are repetitive.
  - Too much pay-to-win: classes and items.
  - Most of the drop items in the game are junk, your bag can be full or garbage very quick.
  - Most of the mobs are not worth killing.
  - Late game progress is terrible since monsters status increase exponentially at that stage.
  - Making a new character is not recommended
  - Auto-status distribution for the character
  
  I believe that this project should offer ways to the server owner customize the server similar how Rathena does for certain features. The code should be modifiable and flexible in a way that can be easy to get into (I know that I am dreaming) and add your own custom ideas. I really wish to play on a custom balanced server with unique events and features but these should come after the core structure for Melia is done. In conclusion I will say that the officials servers core behaviours/features are worth developing for now because it will bring a much more possibilities of customization later on.

**Comment by Adephx** [comment]
I believe it's perfectly acceptable for us to **NOT** strive for perfection or attempt to emulate the original server 100%. As someone who has played Tree of Savior extensively, I can assure you that 99.99% of the player base wouldn't notice if a skill's damage multiplier was set at 90% or 110%. The key is to ensure that the skill effects match their visual representations. Beyond that, there's room for interpretation.

ToS currently has no active fanbase; it's a dead game, it was a dead game 5 years ago when there were only 110 active players on my EU server.. What Melia should focus on is reviving the game with some artistic freedom.

While I agree that our server emulation should be based on the latest official version for ease of access to information, I strongly suggest that we focus on implementing features retrospectively. Let's compile a list of features that defined the 1.0 build and prioritize those first. Once we have a stable build with 1.0 features, we can start adding everything else for those who prefer the later revisions.

ToS didn't lose its initial qualities; they just got buried under a ton of unnecessary changes over the years.

Here are the features that made ToS great, off the top of my head:

- Fully open player-to-player trading
- No item binding
- Player shops
- Field bosses dropping crafting components
- Mobs dropping crafting components
- Mobs dropping silver
- Class/job synergies
- Challenging Dungeons and Raids
- A robust crafting system 

These features aren't exclusive to the 1.0 build, but most of them are severely limited in the current version. As far as I know, mobs no longer drop silver, player-to-player trading is mostly disabled, mobs no longer drop crafting components, and there are no player shops. 

These changes seem like issues that can solved with an on/off toggle.

I also think, that we should be running a public server for alpha testing, as soon as the server is somewhat feature complete (fully working skills, jobs, mobs, crafting, etc.)

**Reply by exectails** [reply]
Perhaps your examples are intentionally hyperbole, but I just want to reiterate that this is not about whether a skill does 110% or 120% damage. For the most part it will never be about minor differences like that. Rather it's about whether you can start only in Klaipeda, or also in Orsha, whether you have circles or not, whether you can freely assign your stat points or the game does it for you, and so on. _Actual_ differences, where users may go either way, because of differing tastes and opinions.

Also, the first two features in your list are peculiar to me, because as I recall...
- Initially there was no trading at all, then it was added as like a premium feature, and you couldn't trade silver, which made the whole thing somewhat useless.
- Items definitely had trade restrictions, like not being able to trade them or put them into storage, is that not effectively the same as item binding?

**Reply by Adephx** [reply]
You could trade items for items since day 1, but you were limited to 5 items per day or something, without the premium account. Before the introduction of transcendance, you could trade anything, but would lose 1 potential doing so.

**Reply by exectails** [reply]
Looked it up, and perhaps I'm misremembering that it wasn't available at all initially, but it was definitely token only in the beginning. Various posts on the official forum from 2016 confirm as much. And even as a token user you were limited to a certain amount of trades per month. I remember it being a hotly debated issue in the early days. Couple that with the "potential" nonsense, a memory I completely suppressed, as well as not being able to trade silver, and I do have to say once more: I'm perplexed by you calling it "fully open"^^ But I guess that's not actually important. I do agree that a game should obviously have actual, free and open trading.

Edit: Oh, I actually just found one more post that said you _were_ able to freely trade anything at any time at some point in the beta. I don't believe this was the case once I got access to the second closed iToS beta, but I guess you were correct that it was a thing.

**Reply by Adephx** [reply]
It was open enough for me not to be bothered by it. It was 30 trades, not items I think and they stacked. I remember being able to make enough silver to afford multiple tokens, so I never used all of my trades and had something like 3 years of premium time. Also, we used some red shards, the one used for transcendance, as an alternative currency to silver. You could also trade items for silver with a personal shop for a while. Alternative was to log at late hour and put the item on the market for the other guy, you suffered a 10% or so silver penalty, but you could get your item to your friend, if you wanted, without even losing potential. There were ways if you were clever enough. Anyhow, I'm getting tired of these discussions, if the goal of melia is to emulate different stages of ToS, including the negatives, so be it, I'm just saying that we should do two versions instead, one as it is and one unrestricted, because like I said, most of the things wrong with ToS can be more or less unrestricted with a simple toggle or two.

![qbvft81](https://github.com/NoCode-NoLife/melia/assets/17905293/2d8de54b-b4ee-4d23-aa37-20fdce69ed5b)

**Reply by exectails** [reply]
> I'm getting tired of these discussions

I was just curious about why you called something "fully open" that was anything but, but alright o.o

**Reply by Adephx** [reply]
Because it was 8 years ago and my memory is fragmented. It was fully opened compared to what it is now. I could trade anything I wanted, including silver, using my methods.

**Comment by henrygepeng** [comment]
How can I join your discord group? I would like to join this project.

**Reply by Akinue** [reply]
The discord link in the [readme](https://github.com/NoCode-NoLife/Melia?tab=readme-ov-file#links)
https://discord.gg/5sszEgw
